Value and order capabilities for schema indexes

burqen opened the pull request:
Builds on top of https://github.com/neo4j/neo4j/pull/10279

Introduce infrastructure to be able to get exact values and defined result order from index queries.

Implementation of capabilities is limited to single property number indexes and only ascending order.

Commit:
Test real index implementations with NodeValueIndexCursorTest

Commit:
Remove redundant dependency

Commit:
Add IndexCapability and CapableIndexReference

First step toward providing index backed ORDER BY in cypher.
IndexCapability answer the questions:
 - Can I get the actual property values from this index?
 - What ordering of index query results does this index naturally provide?

IndexCapability describe the functional capabilities of a given index.
Each schema index rule stored in schema store will have it's own
capabilities depending on index type (unique, general), index description
(label, properties) and index provider.

Right now IndexCapability is decided by the SchemaIndexProvider
backing the given index.

IndexCapability can not change for a given index during its lifetime.

Commit:
Introduce IndexMeta

as container for:
- IndexDescriptor
- SchemaIndexProvider.Descriptor
- IndexCapability

Those three where passed to IndexProxy
instances individually even though they together
describe an index completely. Now they are contained
in one IndexMeta class instead for simplicity.

Commit:
Implement IndexValueCapability for Native schema number index

Commit:
Introduce IndexOrder capability for schema index queries

Schema indexes report what orderings (ascending, descending)
it is capable of providing for given combination of ValueGroups.

A specific ordering of results can be requested through Read.nodeIndexSeek
or Read.nodeIndexScan. Requested ordering must be supported by given
IndexReference.

At this moment, only single property number indexes are capable
of any result ordering and that ordering is ascending.
Provided by NativeSchemaIndexProvider through FusionSchemaIndexProvider.

Commit:
Test for UnionIndexCapability

Commit:
Missing license

Commit:
Remove old IndexProgressor interface

Commit:
Order test for NativeSchemaNumberIndexReader

klaren commented on community/kernel/src/main/java/org/neo4j/kernel/impl/newapi/UnionIndexCapability.java:
```diff
@@ -0,0 +1,69 @@
+        Set<IndexOrder> orderCapability = new HashSet<>();
+        for ( IndexCapability capability : capabilities )
+        {
+            Arrays.stream( capability.order( valueGroups ) ).forEach( orderCapability::add );
```
```
orderCapability.addAll( Arrays.asList( capability.order( valueGroups ) ) );
```
would be easier to follow.

klaren commented on community/kernel/src/test/java/org/neo4j/kernel/impl/newapi/MockStore.java:
```diff
@@ -25,9 +25,9 @@
 
 import org.neo4j.collection.primitive.Primitive;
 import org.neo4j.collection.primitive.PrimitiveLongObjectMap;
+import org.neo4j.internal.kernel.api.*;
```
Star imports not allowed, I thought that checkstyle caught those.

burqen replied:
It does apparently https://build.neohq.net/viewLog.html?buildId=2383763&buildTypeId=PullRequests_BranchPackaging&tab=buildLog&_focus=4287

klaren commented on community/kernel-api/src/test/java/org/neo4j/internal/kernel/api/NodeValueIndexCursorTestBase.java:
```diff
@@ -85,51 +96,73 @@ public void shouldPerformExactLookup() throws Exception
 
             // then
-            assertNodes( node, uniqueIds, num6 );
+            // todo continue here
```
todo

klaren commented on community/kernel-api/src/test/java/org/neo4j/internal/kernel/api/NodeValueIndexCursorTestBase.java:
```diff
@@ -312,3 +485,4 @@ private long nodeWithProp( GraphDatabaseService graphDb, Object value )
         return node.getId();
     }
 }
+
```
2 new line at end of file

klaren reviewed (dismissed):
Looks really good! I just have a few small comments.

Commit:
Remove done() from Node*Client

Instead, progressor.next() will be used directly as
an indicator if client is done or not.

This means NodeValueClientFilter now sits in the
middle between NodeValueClient and IndexProgressor
and passes calls between them.

fickludd commented on community/kernel-api/src/main/java/org/neo4j/internal/kernel/api/NodeValueIndexCursor.java:
```diff
@@ -23,6 +23,37 @@
+ *     for ( int i = 0; i < nbrOfProps; i++ )
+ *     {
+ *         propertyKeys[i] = cursor.propertyKey( i );
+ *     }
```
Dunno if we actually need the `propertyKeys` array here?

burqen replied:
Tried to illustrate that you can get targeted property ids from cursor and if we are missing `ValueCapability` then we can do 
```
 *             for ( int i = 0; i < nbrOfProps; i++ )
 *             {
 *                 values[i] = getPropertyValueFromStore( cursor.nodeReference(), propertyKeys[i] );
 *             }
```
But that was not very clear apparently 👍 
Thinking about it now it seems to just obscure the pattern that I'm actually trying to communicate.

fickludd replied:
`values[i] = getPropertyValueFromStore( cursor.nodeReference(), cursor.propertyKey( i ) )`?

fickludd commented on community/kernel-api/src/main/java/org/neo4j/internal/kernel/api/IndexValueCapability.java:
```diff
@@ -0,0 +1,47 @@
+public enum IndexValueCapability
+{
+    YES( 3 ),
+    MAYBE( 2 ),
```
Would love to know that `MAYBE` means? For some values? Could be `PARTIAL` instead?

burqen replied:
Exactly. I like `PARTIAL` 👍

fickludd commented on community/kernel-api/src/test/java/org/neo4j/internal/kernel/api/NodeValueIndexCursorTestBase.java:
```diff
@@ -64,6 +69,12 @@ void createTestGraph( GraphDatabaseService graphDb )
             strThree3 = nodeWithProp( graphDb, "three" );
             nodeWithProp( graphDb, false );
             boolTrue = nodeWithProp( graphDb, true );
+            nodeWithProp( graphDb, 3 ); // Purposely mix ordering
```
nice!

fickludd commented on community/kernel-api/src/test/java/org/neo4j/internal/kernel/api/NodeValueIndexCursorTestBase.java:
```diff
@@ -85,51 +96,73 @@ public void shouldPerformExactLookup() throws Exception
         {
             // when
-            read.nodeIndexSeek( index, node, IndexQuery.exact( prop, "zero" ) );
+            IndexValueCapability expectValue = index.value( ValueGroup.TEXT );
```
This line i confusing to me. `index.value( ... )` reads like a query or value-access, not a capability-listing. And is this really an `expectValue`?

burqen replied:
Change this to `IndexValueCapability valueCapability = index.valueCapability( ValueGroup.TEXT );`

fickludd commented on community/kernel-api/src/test/java/org/neo4j/internal/kernel/api/NodeValueIndexCursorTestBase.java:
```diff
@@ -157,15 +191,16 @@ public void shouldPerformStringSuffixSearch() throws Exception
         int prop = token.propertyKey( "prop" );
-        IndexReference index = schemaRead.index( label, prop );
+        CapableIndexReference index = schemaRead.index( label, prop );
+        IndexValueCapability stringCapability = index.value( ValueGroup.TEXT );
```
I like `stringCapability` much better!

fickludd commented on community/kernel-api/src/test/java/org/neo4j/internal/kernel/api/NodeValueIndexCursorTestBase.java:
```diff
@@ -280,15 +311,157 @@ public void shouldPerformIndexScan() throws Exception
         int prop = token.propertyKey( "prop" );
-        IndexReference index = schemaRead.index( label, prop );
+        CapableIndexReference index = schemaRead.index( label, prop );
+        IndexValueCapability wildcardCapability = index.value( new ValueGroup[]{null} );
```
WDYT about passing an empty var-args here instead of `new ValueGroup[]{null}`?

burqen replied:
The reason we want to use an explicit null value in array is because of composite indexes where you can have something like `new ValueGroup[]{null,ValueGroup.TEXT}`, indicating that first part of key could be of any value group and the second part is known to be a string.

Do you still think it is reasonable to use empty array for single property wildcard? I can see the use for that syntactic sugar but at the same time I don't like multiple ways of doing the same thing.

I guess another alternative would be to add another `ValueGroup.WILDCARD` or use `ValueGroup.NO_VALUE` for that purpose. But doesn't seem right to pollute `Values` module like that.

fickludd replied:
We have a `ValueGroup.NO_VALUE` already. Maybe we should add `ValueGroup.ANY`?

burqen replied:
I'm gonna try that 🎉

fickludd commented on community/kernel-api/src/test/java/org/neo4j/internal/kernel/api/NodeValueIndexCursorTestBase.java:
```diff
@@ -280,15 +311,157 @@ public void shouldPerformIndexScan() throws Exception
         {
             // when
-            read.nodeIndexScan( index, node );
+            read.nodeIndexScan( index, node, null );
```
Also we could add a `IndexOrder.ARBITRARY` or `ANY_ORDER` or something instead of passing `null`.

burqen replied:
We want `IndexOrder` to be used both as the answer to "what different ordering are you capable of?" and as the means of requesting order, "I want order X for my result." That makes interpretation of `ARBITRARY` and `ANY_ORDER` a bit fuzzy.

May `UNORDERED` could be a good fit.

fickludd replied:
Hmm, `read.nodeIndexScan( index, node, UNORDERED )` reads like we explicitly want the results to be out of order. Maybe overload the method `read.nodeIndexScan( index, node )`? Or `read.nodeIndexScan( index, node, ANY_ORDER )` would read ok, but that crashes with the capability use. 

Meh. Maybe keep the null then :)

burqen replied:
@MishaDemianenko Please read this discussion regarding passing `null` as `IndexOrder`.
We flipped back and forth and landed on `null` in the end, but I'm willing to go another lap ;)

fickludd commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/NumberHitIndexCursorProgressor.java:
```diff
@@ -23,61 +23,59 @@
 {
     private final RawCursor<Hit<KEY,VALUE>,IOException> seeker;
-    private final Collection<RawCursor<Hit<KEY,VALUE>,IOException>> toRemoveFromWhenExhausted;
+    private final NodeValueClient cursor;
```
`client` now, post-rebase =)

fickludd commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexReader.java:
```diff
@@ -100,6 +107,42 @@ public PrimitiveLongIterator query( IndexQuery... predicates ) throws IndexNotAp
+            return;
+        }
+
+        // todo: There will be no ordering of the node ids here. Is this a problem?
```
If we provide an `ExistsPredicate` with a specified `IndexOrder` we need to fail if we cannot serve it?

fickludd commented on community/kernel/src/main/java/org/neo4j/kernel/impl/newapi/IndexReference.java:
```diff
@@ -0,0 +1,71 @@
+{
+    private final int label;
+    private final int[] properties;
+    private final boolean unique;
```
It somehow feels wrong to create a new index meta class that is a `IndexDescriptor` + capabilities. Could we do composition instead? We don't have to fix it now, just venting opinion.

burqen replied:
Yes, I agree it's a bit awkward. From what we discussed yesterday we might want to split up the code path in two.
```
public interface SchemaRead
{
    IndexDescription index( int label, int... properties );

    IndexCapability indexCapability( IndexDescription indexDescription );
}
```
Let's address this issue at that point.

fickludd commented on community/kernel/src/test/java/org/neo4j/kernel/impl/newapi/ReadTestSupport.java:
```diff
@@ -21,22 +21,34 @@
+    void addSetting( Setting setting, String value )
+    {
+        settings.put( setting, value );
+    }
```
This is a nice addition!

fickludd commented on community/kernel/src/test/java/org/neo4j/kernel/impl/newapi/SimpleNodeValueClient.java:
```diff
@@ -0,0 +1,57 @@
+import org.neo4j.storageengine.api.schema.IndexProgressor;
+import org.neo4j.values.storable.Value;
+
+public class SimpleNodeValueClient implements IndexProgressor.NodeValueClient
```
As an alternative, this could live together with `IndexProgressor` in `org.neo4j.storageengine.api.schema`

burqen replied:
I don't see any use for this class outside of tests, so let's keep it out of production code.

fickludd replied:
Yes... I'm suggesting to put it in `community/kernel/src/test/org/neo4j/storageengine/api/schema`?

burqen replied:
Haha, alright. Sounds good!

fickludd reviewed (changes requested):
This is great. Got some opportunities where readability could improve, but only nitpicks.

Commit:
Remove star imports

Commit:
Clarify naming around index capabilities

Commit:
Rename NumberHitIndexCursorProgressor -> NumberHitIndexProgressor

Commit:
FusionIndexReader validate index order on queries

burqen commented:
@fickludd @klaren Really sweet feedback! :D I think I've addressed all of your comments now.

Commit:
Checkstyle

Commit:
Move SimpleNodeValueClient to org.neo4j.storageengine.api.schema

Commit:
Use ValueGroup.UNKNOWN as wildcard instead of null

Commit:
Remove explicit IndexOrder.UNORDERED and use null instead

burqen commented:
@fickludd Another attempt 😄

Commit:
Small javadoc change

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/api/schema/index/IndexDescriptor.java:
```diff
@@ -41,7 +40,7 @@
  * This will be renamed to IndexDescriptor, once the old org.neo4j.kernel.api.schema.IndexDescriptor is completely
```
since you already here can you please clean up this comment

fickludd replied:
Hehe, whoops!

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionSchemaIndexProvider.java:
```diff
@@ -126,6 +130,27 @@ public InternalIndexState getInitialState( long indexId, IndexDescriptor descrip
+            public IndexOrder[] orderCapability( ValueGroup... valueGroups )
+            {
+                // No order capability when combining results from different indexes
+                if ( valueGroups.length == 1 && valueGroups[0] == ValueGroup.UNKNOWN )
```
shouldn't we have the same check for value capabilities?

burqen replied:
We need fusion to make this decision about ordering because he's responsible for combining the results and thus affect the ordering. The value capability can be left completely to underlying implementations and are combined with `UnionIndexCapability`.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/NativeSchemaNumberIndexProvider.java:
```diff
@@ -193,4 +204,41 @@ public boolean compatibleWith( long layoutIdentifier, int majorVersion, int mino
+        {
+            if ( support( valueGroups ) )
+            {
+                return new IndexOrder[]{IndexOrder.ASCENDING};
```
we do not really need new arrays each time here right?

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/NativeSchemaNumberIndexProvider.java:
```diff
@@ -193,4 +204,41 @@ public boolean compatibleWith( long layoutIdentifier, int majorVersion, int mino
+
+        private boolean support( ValueGroup[] valueGroups )
+        {
+            return valueGroups.length == 1 && ValueGroup.NUMBER.equals( valueGroups[0] );
```
please use == for enum comparisons like its done 2 lines above

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/NativeSchemaNumberIndexReader.java:
```diff
@@ -106,37 +113,64 @@ public long countIndexedNodes( long nodeId, Value... propertyValues )
     public PrimitiveLongIterator query( IndexQuery... predicates ) throws IndexNotApplicableKernelException
     {
+        NodeValueIterator nodeValueIterator = new NodeValueIterator();
+        query( nodeValueIterator, null, predicates );
```
can we avoid passing null here? and introduce constant like IndexOrder.NONE for that instead?

burqen replied:
Please read thread above with @fickludd . (I'm absolutely willing to consider this again)

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/NodeValueIterator.java:
```diff
@@ -0,0 +1,71 @@
+public class NodeValueIterator extends PrimitiveLongCollections.PrimitiveLongBaseIterator
+        implements IndexProgressor.NodeValueClient, Resource
+{
+    private volatile boolean closed;
```
why do we need this field to be volatile? do we expect multithreaded close calls?

burqen replied:
Nope. Legacy. Nice catch!

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/NodeValueIterator.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    @Override
+    public void close()
```
NodeValueIterator iterator used only in NativeSchemaNumberIndexReader as PrimitiveLongIterator which is not Resource how this can be reliably closed?

burqen replied:
I'm working on a PR that will make `IndexReader` return `PrimitiveLongResourceIterator` instead. But right now we rely on this snippet from `NodeValueIndexProgressor`
```
    @Override
    public void close()
    {
        if ( ids instanceof Resource )
        {
            ((Resource) ids).close();
        }
    }
```

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/newapi/IndexReference.java:
```diff
@@ -0,0 +1,71 @@
+import org.neo4j.internal.kernel.api.IndexValueCapability;
+import org.neo4j.values.storable.ValueGroup;
+
+class IndexReference implements org.neo4j.internal.kernel.api.CapableIndexReference
```
we have IndexReference class that implements IndexReference interface would be nicer to rename first one

burqen replied:
Suggestions: `IndexHandle`, `CapableIndexReference` (same as interface), `IndexReferenceImpl`, ... hmm... do you have any suggestions?

MishaDemianenko commented on community/kernel-api/src/main/java/org/neo4j/internal/kernel/api/IndexQuery.java:
```diff
@@ -175,6 +176,13 @@ public boolean acceptsValueAt( PropertyCursor property )
     }
 
+    /**
+     * todo Is this useful?
```
not sure not sure
also it looks like we never return null from here

MishaDemianenko commented on community/kernel-api/src/main/java/org/neo4j/internal/kernel/api/IndexValueCapability.java:
```diff
@@ -0,0 +1,47 @@
+ */
+package org.neo4j.internal.kernel.api;
+
+public enum IndexValueCapability
```
would be really nice to get a description what it is about, also what those constants means (maybe having description about value capability in general will be enough)

MishaDemianenko commented on community/kernel-api/src/main/java/org/neo4j/internal/kernel/api/IndexValueCapability.java:
```diff
@@ -0,0 +1,47 @@
+    /**
+     * Higher order indicate a higher capability.
+     */
+    int order;
```
private

MishaDemianenko commented on community/kernel-api/src/main/java/org/neo4j/internal/kernel/api/IndexValueCapability.java:
```diff
@@ -0,0 +1,47 @@
+     */
+    public int compare( IndexValueCapability other )
+    {
+        return order - other.order;
```
Integer.compare

MishaDemianenko reviewed (changes requested):
A couple of comments here and there.

Also as a side note, i think we need to do something with tons of indexes related value objects that exist there in various reincarnations: descriptors, meta, references - it's too much they all do almost the same things, we only make api and reading experience more and more complex

Commit:
Small cleanup

burqen commented:
@MishaDemianenko Thanks for the feedback! I've addressed all comments except for
 - `IndexReference` class name
 - Using `null` as `IndexOrder`

Please come back to me about those 😄

Commit:
Rename IndexReference implementation to DefaultCapableIndexReference

Commit:
Use IndexOrder.NONE instead of null to indicate no ordering required

burqen commented:
@MishaDemianenko Last two issues are now addessed

Commit:
Move native index reader test to new query api
